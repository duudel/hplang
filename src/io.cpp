
#include "io.h"
#include <cstdio>

namespace hplang
{

    s64 PrintString(IoFile *file, String str)
    {
        return fwrite(str.data, 1, str.size, (FILE*)file);
    }

} // hplang
