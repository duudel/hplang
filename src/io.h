#ifndef H_HPLANG_IO_H

#include "types.h"

namespace hplang
{

    // Type punning of c-standard FILE. The dummies who specified what FILE is
    // did not specify it would be struct, so now it is a typedef of an unknown
    // type in most implementations. This makes it impossible to forward
    // declare FILE in the headers, forcing to include stdio. Unless... we use
    // type punning... which is not the most elegant solution.
    struct IoFile;

    s64 PrintString(IoFile *file, String str);

    inline s64 PrintName(IoFile *file, Name name)
    { return PrintString(file, name.str); }

} //hplang

#define H_HPLANG_IO_H
#endif
