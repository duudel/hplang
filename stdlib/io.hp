// io module
// 2016-04-19

foreign {
    FILE :: struct { }
    hp_get_stdout :: () : FILE*;
    hp_get_stderr :: () : FILE*;
    hp_get_stdin :: () : FILE*;

    hp_fwrite :: (file : FILE*, size : s64, data : u8*) : s64;
    hp_fprint_uint :: (file : FILE*, x : u64) : s64;
    hp_fprint_int :: (file : FILE*, x : s64) : s64;
    hp_fprint_f32 :: (file : FILE*, x : f32, precision : s32) : s64;
    hp_fprint_f64 :: (file : FILE*, x : f64, precision : s32) : s64;
}

stdout : FILE* = hp_get_stdout();
stderr : FILE* = hp_get_stderr();
stdin : FILE* = hp_get_stdin();


print :: (s : string) : s64
{ return hp_fwrite(stdout, s.size, s.data -> u8*); }

println :: (s : string) : s64
{
    n := print(s);
    return n + print("\n");
}

println :: () : s64
{ return print("\n"); }

// u64

print :: (x : u64) : s64
{ return hp_fprint_uint(stdout, x); }

println :: (x : u64) : s64
{
    n := hp_fprint_uint(stdout, x);
    return n + println();
}

// s64

// NOTE: This version written in hplang is here only for language robustness
// testing purposes.
print :: (x : s64) : s64
{
    if (x == 0)  return print("0");
    neg := x < 0;
    x = (neg ? -x : x);

    val := x;
    magnitude : s64 = 1;
    while (val > 0)
    {
        val /= 10;
        magnitude *= 10;
    }
    magnitude /= 10;

    written : s64 = 0;
    if neg  written += print("-");
    while (magnitude > 0)
    {
        y := x / magnitude;
        x -= y * magnitude;
        magnitude /= 10;
        //hp_fprint_int(stdout, y);
        //print("; ");
        //c : u8 = (('0' -> u8) + y) -> u8;
        c := ('0' -> u8);
        c += y -> u8;
        written += hp_fwrite(stdout, 1, &c);
    }
    return written;
    //return hp_fprint_int(stdout, x);
}

println :: (x : s64) : s64
{
    n := print(x);
    return n + println();
}

// f32

print :: (x : f32) : s64
{ return hp_fprint_f32(stdout, x, 6); }

print :: (x : f32, precision : s32) : s64
{ return hp_fprint_f32(stdout, x, precision); }

println :: (x : f32) : s64
{
    n := hp_fprint_f32(stdout, x, 6);
    return n + println();
}

println :: (x : f32, precision : s32) : s64
{
    n := hp_fprint_f32(stdout, x, precision);
    return n + println();
}

// f64

print :: (x : f64) : s64
{ return hp_fprint_f64(stdout, x, 6); }

print :: (x : f64, precision : s32) : s64
{ return hp_fprint_f64(stdout, x, precision); }

println :: (x : f64) : s64
{
    n := hp_fprint_f64(stdout, x, 6);
    return n + println();
}

println :: (x : f64, precision : s32) : s64
{
    n := hp_fprint_f64(stdout, x, precision);
    return n + println();
}

