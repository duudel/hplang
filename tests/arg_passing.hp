// Tests argument passing
// 2016-05-19

My_Struct :: struct
{
    bval : bool;
    fval : f64;
}

func :: (a : s32, b : f64, c : My_Struct, d : bool, e : My_Struct)
{
    x := c.bval ? 10 : 16;
    e.fval = 0.5;
}

main :: ()
{
    s1 : My_Struct;
    s2 : My_Struct;
    func(15, 12.5, s1, false, s2);

    return 0;
}
