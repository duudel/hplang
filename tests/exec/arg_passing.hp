// Tests argument passing
// 2016-05-19

import ":io";

My_Struct :: struct
{
    bval : bool;
    fval : f64;
    ival : s16;
}

func ::
    ( a : s32           // 1.
    , b : f64           // 2.
    , c : My_Struct     // 3.
    , d : bool          // 4.
    , e : My_Struct     // 5.
    , f : bool          // 6.
    , g : s32           // 7.
    , h : u8            // 8.
    , i : My_Struct)    // 9.
{
    x := c.bval ? 10 : 16;
    e.fval = 0.5;
    i.ival += (-7) -> s16;
    return x * c.ival + i.ival - h;
}

main :: ()
{
    s1 : My_Struct;
    s2 : My_Struct;

    s1.bval = false;
    s1.fval = 15.0;
    s1.ival = 8 -> s16;

    s2.bval = true;
    s2.fval = 0.0;
    s2.ival = 12 -> s16;

    f := true;
    g := -5;
    h := 2;

    s3 : My_Struct;
    s3.bval = false;
    s3.fval = 128.24;
    s3.ival = 1 -> s16;

    result := func(15, 12.5, s1, false, s2, f, g, h -> u8, s3);

    s2.bval = true;
    print("s2.fval = ");
    print(s2.fval);
    println();
    print("s3.ival = ");
    println(s3.ival);

    return result;
}
