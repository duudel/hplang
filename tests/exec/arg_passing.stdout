s2.fval = 0.500000
s3.ival = -6
